pkg/retrieval/retrieval.go

```diff
@@ -108,51 +123,87 @@ func (s *Service) RetrieveChunk(ctx context.Context, addr swarm.Address) (swarm.
 		var (
 			peerAttempt  int
 			peersResults int
-			resultC      = make(chan swarm.Chunk, maxPeers)
-			errC         = make(chan error, maxPeers)
+			resultC      = make(chan retrievalResult, maxSelects)
 		)
 
-		for {
-			if peerAttempt < maxPeers {
-				peerAttempt++
+		requestAttempt := 0
 
-				s.metrics.PeerRequestCounter.Inc()
+		lastTime := time.Now().Unix()
 
-				go func() {
-					chunk, peer, err := s.retrieveChunk(ctx, addr, sp)
-					if err != nil {
-						if !peer.IsZero() {
-							logger.Debugf("retrieval: failed to get chunk %s from peer %s: %v", addr, peer, err)
-						}
+		for requestAttempt < 5 {
 
-						errC <- err
-						return
-					}
+			if peerAttempt < maxSelects {
 
-					resultC <- chunk
+				span, _, ctx := s.tracer.StartSpanFromContext(context.Background(), "retrieve-chunk", s.logger, opentracing.Tag{Key: "address", Value: addr.String()})
+				defer span.Finish()
+
+				peerAttempt++
+				s.metrics.PeerRequestCounter.Inc()
+				go func() {
+					chunk, peer, requested, err := s.retrieveChunk(ctx, addr, sp)
+					resultC <- retrievalResult{
+						chunk:     chunk,
+						peer:      peer,
+						err:       err,
+						retrieved: requested,
+					}
 				}()
 			} else {
-				ticker.Stop()
+				resultC <- retrievalResult{}
 			}
 
 			select {
 			case <-ticker.C:
 				// break
-			case chunk := <-resultC:
-				return chunk, nil
-			case <-errC:
-				peersResults++
+			case res := <-resultC:
+				if res.retrieved {
+					if res.err != nil {
+						if !res.peer.IsZero() {
+							s.logger.Debugf("retrieval: failed to get chunk %s from peer %s: %v", addr, res.peer, res.err)
+						}
+						peersResults++
+					} else {
+						return res.chunk, nil
+					}
+				}
 			case <-ctx.Done():
-				logger.Tracef("retrieval: failed to get chunk %s: %v", addr, ctx.Err())
+				s.logger.Tracef("retrieval: failed to get chunk %s: %v", addr, ctx.Err())
 				return nil, fmt.Errorf("retrieval: %w", ctx.Err())
 			}
 
-			// all results received
+			// all results received, only successfully attempted requests are counted
 			if peersResults >= maxPeers {
-				logger.Tracef("retrieval: failed to get chunk %s", addr)
+				s.logger.Tracef("retrieval: failed to get chunk %s", addr)
 				return nil, storage.ErrNotFound
 			}
+
+			// if we have not counted enough successful attempts but out of selection amount, reset
+			if peerAttempt >= maxSelects {
+				if !origin {
+					return nil, storage.ErrNotFound
+				}
+
+				requestAttempt++
+				timeNow := time.Now().Unix()
+				if timeNow > lastTime {
```

Comment:
This would always be true, no?

Reply:
if we are in the same second still
we wait 600 miliseconds only so we should get into the next second within 2 rounds of requestAttempts

Reply:
Ok, can we add a test case that covers the area starting at `requestAttempt++` and down.

Reply:
I leave the judgement to you if this is necessary